README.md
# confetti-react

Confetti without the cleanup. [Demo](http://rkuykendall.github.io/confetti-react/)

[![Build Status](https://travis-ci.org/rkuykendall/confetti-react.svg?branch=master)](https://travis-ci.org/rkuykendall/confetti-react)
[![npm](https://img.shields.io/npm/v/confetti-react.svg)](https://www.npmjs.com/package/confetti-react)
[![npm bundle size](https://img.shields.io/bundlephobia/min/confetti-react.svg)](https://bundlephobia.com/result?p=confetti-react)
![npm type definitions](https://img.shields.io/npm/types/confetti-react.svg)

Based on a pen by @Gthibaud: https://codepen.io/Gthibaud/pen/ENzXbp

[![demogif][2]][1]

[1]: http://rkuykendall.github.com/confetti-react
[2]: http://rkuykendall.github.io/confetti-react/confetti-demo-anim.gif 'demo gif'

## Install

```sh
npm install confetti-react
```

## Use

`width` and `height` props are recommended. They will default to the initial window dimensions, but will not respond to
resize events. It is recommended to provide the dimensions yourself. Here is an example using
[a hook](https://github.com/streamich/react-use/blob/master/docs/useWindowSize.md):

```jsx
import React from 'react';
import useWindowSize from 'react-use/lib/useWindowSize';
import Confetti from 'confetti-react';

export default () => {
  const { width, height } = useWindowSize();
  return <Confetti width={width} height={height} />;
};
```

## Props

| Property             | Type                                                                                                       | Default                                                                                                                                                                                                                                                                            | Description                                                           |
| -------------------- | ---------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------- |
| `width`              | `Number`                                                                                                   | `window.innerWidth \|\| 300`                                                                                                                                                                                                                                                       | Width of the `<canvas>` element.                                      |
| `height`             | `Number`                                                                                                   | `window.innerHeight \|\| 200`                                                                                                                                                                                                                                                      | Height of the `<canvas>` element.                                     |
| `numberOfPieces`     | `Number`                                                                                                   | 200                                                                                                                                                                                                                                                                                | Number of confetti pieces at one time.                                |
| `confettiSource`     | `{ x: Number, y: Number, w: Number, h: Number }`                                                           | `{x: 0, y: 0, w: canvas.width, h:0}`                                                                                                                                                                                                                                               | Rectangle where the confetti should spawn. Default is across the top. |
| `friction`           | `Number`                                                                                                   | 0.99                                                                                                                                                                                                                                                                               |                                                                       |
| `wind`               | `Number`                                                                                                   | 0                                                                                                                                                                                                                                                                                  |                                                                       |
| `gravity`            | `Number`                                                                                                   | 0.1                                                                                                                                                                                                                                                                                |                                                                       |
| `initialVelocityX`   | `Number`                                                                                                   | 4                                                                                                                                                                                                                                                                                  | How fast confetti is emitted horizontally                             |
| `initialVelocityY`   | `Number`                                                                                                   | 10                                                                                                                                                                                                                                                                                 | How fast confetti is emitted vertically                               |
| `colors`             | `String[]`                                                                                                 | `['#f44336'`</br>`'#e91e63'`</br>`'#9c27b0'`</br>`'#673ab7'`</br>`'#3f51b5'`</br>`'#2196f3'`</br>`'#03a9f4'`</br>`'#00bcd4'`</br>`'#009688'`</br>`'#4CAF50'`</br>`'#8BC34A'`</br>`'#CDDC39'`</br>`'#FFEB3B'`</br>`'#FFC107'`</br>`'#FF9800'`</br>`'#FF5722'`</br>`'#795548']`</br> | All available Colors for the confetti pieces.                         |
| `opacity`            | `Number`                                                                                                   | 1.0                                                                                                                                                                                                                                                                                |                                                                       |
| `recycle`            | `Bool`                                                                                                     | true                                                                                                                                                                                                                                                                               | Keep spawning confetti after `numberOfPieces` pieces have been shown. |
| `run`                | `Bool`                                                                                                     | true                                                                                                                                                                                                                                                                               | Run the animation loop                                                |
| `tweenDuration`      | `Number`                                                                                                   | 5000                                                                                                                                                                                                                                                                               | How fast the confetti is added                                        |
| `tweenFunction`      | `(currentTime: number, currentValue: number, targetValue: number, duration: number, s?: number) => number` | easeInOutQuad                                                                                                                                                                                                                                                                      | See [tween-functions](https://github.com/chenglou/tween-functions)    |
| `drawShape`          | `(context: CanvasRenderingContext2D) => void`                                                              | `undefined`                                                                                                                                                                                                                                                                        | See below                                                             |
| `onConfettiComplete` | `(confetti: Confetti) => void`                                                                             | `undefined`                                                                                                                                                                                                                                                                        | Called when all confetti has fallen off-canvas.                       |

# `drawShape()`

Draw a custom shape for a particle. If not provided, defaults to a random selection of a square, circle or strip
confetto. The function is called with the canvas context as a parameter and the [Particle](src/Particle.ts) as the
`this` context.

For example, to draw all spirals:

```jsx
<Confetti
  drawShape={ctx => {
    ctx.beginPath();
    for (let i = 0; i < 22; i++) {
      const angle = 0.35 * i;
      const x = (0.2 + 1.5 * angle) * Math.cos(angle);
      const y = (0.2 + 1.5 * angle) * Math.sin(angle);
      ctx.lineTo(x, y);
    }
    ctx.stroke();
    ctx.closePath();
  }}
/>
```


CHANGELOG.md
## [2.4.1](https://github.com/rkuykendall/confetti-react/compare/v2.4.0...v2.4.1) (2019-02-28)

### Bug Fixes

- **test:** testing semantic-release ([ef3dfa8](https://github.com/rkuykendall/confetti-react/commit/ef3dfa8))


src/Confetti.ts
import tweens from 'tween-functions';
import { Rect } from './Rect';
import ParticleGeneratorClass from './ParticleGeneratorClass';

export interface ConfettiOptions {
  /**
   * Width of the component
   * @default window.width
   */
  width: number;
  /**
   * Height of the component
   * @default window.height
   */
  height: number;
  /**
   * Max number of confetti pieces to render.
   * @default 200
   */
  numberOfPieces: number;
  /**
   * Slows movement of pieces. (lower number = slower confetti)
   * @default 0.99
   */
  friction: number;
  /**
   * Blows confetti along the X axis.
   * @default 0
   */
  wind: number;
  /**
   * How fast it falls (pixels per frame)
   * @default 0.1
   */
  gravity: number;
  /**
   * How fast the confetti is emitted horizontally
   * @default 4
   */
  initialVelocityX: number;
  /**
   * How fast the confetti is emitted vertically
   * @default 10
   */
  initialVelocityY: number;
  /**
   * Array of colors to choose from.
   */
  colors: string[];
  /**
   * Opacity of the confetti.
   * @default 1
   */
  opacity: number;
  /**
   * If false, only numberOfPieces will be emitted and then stops. If true, when a confetto goes offscreen, a new one will be emitted.
   * @default true
   */
  recycle: boolean;
  /**
   * If false, stops the requestAnimationFrame loop.
   * @default true
   */
  run: boolean;
  /**
   * Renders some debug text on the canvas.
   * @default false
   */
  debug: boolean;
  /**
   * A Rect defining the area where the confetti will spawn.
   * @default {
   *   x: 0,
   *   y: 0,
   *   w: canvas.width,
   *   h: 0
   * }
   */
  confettiSource: Rect;
  /**
   * Controls the rate at which confetti is spawned.
   * @default easeInOutQuad
   */
  tweenFunction: (
    currentTime: number,
    currentValue: number,
    targetValue: number,
    duration: number,
    s?: number,
  ) => number;
  /**
   * Number of milliseconds it should take to spawn numberOfPieces.
   * @default 5000
   */
  tweenDuration: number;
  /**
   * Function to draw your own confetti shapes.
   */
  drawShape?: (context: CanvasRenderingContext2D) => void;
  /**
   * Function called when all confetti has fallen off-canvas.
   */
  onConfettiComplete?: (confettiInstance?: Confetti) => void;
}

export const confettiDefaults: Pick<ConfettiOptions, Exclude<keyof ConfettiOptions, 'confettiSource'>> = {
  width: typeof window !== 'undefined' ? window.innerWidth : 300,
  height: typeof window !== 'undefined' ? window.innerHeight : 200,
  numberOfPieces: 200,
  friction: 0.99,
  wind: 0,
  gravity: 0.1,
  initialVelocityX: 4,
  initialVelocityY: 10,
  colors: [
    '#f44336',
    '#e91e63',
    '#9c27b0',
    '#673ab7',
    '#3f51b5',
    '#2196f3',
    '#03a9f4',
    '#00bcd4',
    '#009688',
    '#4CAF50',
    '#8BC34A',
    '#CDDC39',
    '#FFEB3B',
    '#FFC107',
    '#FF9800',
    '#FF5722',
    '#795548',
  ],
  opacity: 1.0,
  debug: false,
  tweenFunction: tweens.easeInOutQuad,
  tweenDuration: 5000,
  recycle: true,
  run: true,
};

export class Confetti {
  constructor(canvas: HTMLCanvasElement, opts: Partial<ConfettiOptions>) {
    this.canvas = canvas;
    const ctx = this.canvas.getContext('2d');
    if (!ctx) {
      throw new Error('Could not get canvas context');
    }
    this.context = ctx;

    this.generator = new ParticleGeneratorClass(this.canvas, () => this.options as ConfettiOptions);
    this.options = opts;
    this.update();
  }

  canvas: HTMLCanvasElement;

  context: CanvasRenderingContext2D;

  _options!: ConfettiOptions;

  generator: ParticleGeneratorClass;

  rafId?: number;

  get options(): Partial<ConfettiOptions> {
    return this._options;
  }

  set options(opts: Partial<ConfettiOptions>) {
    const lastRunState = this._options && this._options.run;
    const lastRecycleState = this._options && this._options.recycle;
    this.setOptionsWithDefaults(opts);
    if (this.generator) {
      Object.assign(this.generator, this.options.confettiSource);
      if (typeof opts.recycle === 'boolean' && opts.recycle && lastRecycleState === false) {
        this.generator.lastNumberOfPieces = this.generator.particles.length;
      }
    }
    if (typeof opts.run === 'boolean' && opts.run && lastRunState === false) {
      this.update();
    }
  }

  setOptionsWithDefaults = (opts: Partial<ConfettiOptions>) => {
    const computedConfettiDefaults = {
      confettiSource: {
        x: 0,
        y: 0,
        w: this.canvas.width,
        h: 0,
      },
    };
    this._options = {
      ...computedConfettiDefaults,
      ...confettiDefaults,
      ...opts,
    };
    Object.assign(this, opts.confettiSource);
  };

  update = () => {
    const {
      options: { run, onConfettiComplete },
      canvas,
      context,
    } = this;
    if (run) {
      context.fillStyle = 'white';
      context.clearRect(0, 0, canvas.width, canvas.height);
    }
    if (this.generator.animate()) {
      this.rafId = requestAnimationFrame(this.update);
    } else {
      if (onConfettiComplete && typeof onConfettiComplete === 'function' && this.generator.particlesGenerated > 0) {
        onConfettiComplete.call(this, this);
      }
      this._options.run = false;
    }
  };

  reset = () => {
    if (this.generator && this.generator.particlesGenerated > 0) {
      this.generator.particlesGenerated = 0;
      this.generator.particles = [];
      this.generator.lastNumberOfPieces = 0;
    }
  };

  stop = () => {
    this.options = { run: false };
    if (this.rafId) {
      cancelAnimationFrame(this.rafId);
      this.rafId = undefined;
    }
  };
}

export default Confetti;
